linera-client/src/wallet.rs

```diff
@@ -0,0 +1,327 @@
+            .filter_map(|user_chain| user_chain.key_pair.as_ref())
+            .find(|key_pair| key_pair.public() == *key)
+            .map(|key_pair| key_pair.copy())
+    }
+
+    pub fn assign_new_chain_to_key(
+        &mut self,
+        key: PublicKey,
+        chain_id: ChainId,
+        timestamp: Timestamp,
+    ) -> Result<(), anyhow::Error> {
+        let key_pair = self
+            .unassigned_key_pairs
+            .remove(&key)
+            .context("could not assign chain to key as unassigned key was not found")?;
+        let user_chain = UserChain {
+            chain_id,
+            key_pair: Some(key_pair),
+            block_hash: None,
+            timestamp,
+            next_block_height: BlockHeight(0),
+            pending_block: None,
+            pending_blobs: BTreeMap::new(),
+        };
+        self.insert(user_chain);
+        Ok(())
+    }
+
+    pub fn set_default_chain(&mut self, chain_id: ChainId) -> Result<(), anyhow::Error> {
+        anyhow::ensure!(
+            self.chains.contains_key(&chain_id),
+            "Chain {} cannot be assigned as the default chain since it does not exist in the \
+             wallet.",
+            &chain_id
+        );
+        self.default = Some(chain_id);
+        Ok(())
+    }
+
+    pub async fn update_from_state<P, S>(&mut self, chain_client: &ChainClient<P, S>)
+    where
+        P: ValidatorNodeProvider + Sync + 'static,
+        S: Storage + Clone + Send + Sync + 'static,
+        ViewError: From<S::StoreError>,
+    {
+        let key_pair = chain_client.key_pair().await.map(|k| k.copy()).ok();
+        let state = chain_client.state();
+        self.chains.insert(
+            chain_client.chain_id(),
+            UserChain {
+                chain_id: chain_client.chain_id(),
+                key_pair,
+                block_hash: state.block_hash,
+                next_block_height: state.next_block_height,
+                timestamp: state.timestamp,
+                pending_block: state.pending_block.clone(),
+                pending_blobs: state.pending_blobs.clone(),
+            },
+        );
+    }
+
+    pub fn genesis_admin_chain(&self) -> ChainId {
+        self.genesis_config.admin_id
+    }
+
+    pub fn genesis_config(&self) -> &GenesisConfig {
+        &self.genesis_config
+    }
+
+    pub fn make_prng(&self) -> Box<dyn CryptoRng> {
+        self.testing_prng_seed.into()
+    }
+
+    pub fn refresh_prng_seed<R: CryptoRng>(&mut self, rng: &mut R) {
+        if self.testing_prng_seed.is_some() {
+            self.testing_prng_seed = Some(rng.gen());
+        }
+    }
+
+    pub fn pretty_print(&self, chain_id: Option<ChainId>) {
+        let mut table = Table::new();
+        table
+            .load_preset(UTF8_FULL)
+            .apply_modifier(UTF8_ROUND_CORNERS)
+            .set_content_arrangement(ContentArrangement::Dynamic)
+            .set_header(vec![
+                Cell::new("Chain Id").add_attribute(Attribute::Bold),
```

Comment:
Is the pretty printing logic duplicated now?

```suggestion
                Cell::new("Chain ID").add_attribute(Attribute::Bold),
```

Reply:
Oh it definitely shouldn't be!  Let me fix that up.